File: javascript/4.js
/*
 * https://projecteuler.net/problem=4
 *
 * A palindromic number reads the same both ways.
 * The largest palindrome made from the product
 * of two 2-digit numbers is 9009 = 91 × 99.
 *
 * Find the largest palindrome made from the product of two 3-digit numbers.
 */

// convert a number into an array of digits
function digitize(num) {
  let digits = [];

  while (num > 0) {
    digits.push(num % 10);
    num = Math.floor(num / 10);
  }

  return digits;
}

// determine if a number is a palendrome
function isPalendrome(num) {
  // digitize the number
  const digits = digitize(num);

  // get the digits in reverse
  const reversed = digits.slice().reverse();

  // compare the digits with the reversed digits
  let result = true;

  for (let i = 0; i < digits.length && result; i++) {
    result = (digits[i] === reversed[i]);
  }

  return result;
}

// solve the largest palendrome problem
function largestPalendrome() {

  // start by assuming the largest found is 0
  let largest = 0;

  // for each 3 digit number
  for (let i = 999; i > 99; i--) {

    /*
     * for every other 3 digit number that produces
     * a product with the first 3 digit number that's
     * larger than the largest palendrome found
     */
    for (let j = 999; j > 99 && i * j > largest; j--) {

      /*
       * check if the product is a palendrome,
       * if so it's the new largest palendrome found
       */
      if (isPalendrome(i * j)) {
        largest = i * j;
      }
    }
  }

  return largest;
}

console.log(largestPalendrome());

File: dart/4.dart
// https://projecteuler.net/problem=4
// 
// A palindromic number reads the same both ways.
// The largest palindrome made from the product
// of two 2-digit numbers is 9009 = 91 x 99.
// 
// Find the largest palindrome made from the product of two 3-digit numbers.

// Convert a number into a list of digits.
List<int> digitize(int n) {
  var digits = <int>[];

  while (n > 0) {
    digits.add(n % 10);
    n ~/= 10;
  }

  return digits;
}

// Determine if a number is a palendrome.
bool isPalendrome(n) {
  // Digitize the number.
  var digits = digitize(n);

  // Get the digits in reversed order.
  var reversed = [...digits.reversed];

  // Compare the digits with the reversed digits.
  var result = true;

  for (var i = 0; i < digits.length && result; i++) {
    result = (digits[i] == reversed[i]);
  }

  return result;
}

// Solve the largest palendrome problem.
void main() {
  // Start by assuming the largest found is 0.
  var largest = 0;

  // For each 3 digit number.
  for (var i = 999; i > 99; i--) {
    for (var j = 999; (j > 99) && (i * j > largest); j--) {
      // Check if the product is a palendrome,
      // if so it's the new largest palendrome found.
      if (isPalendrome(i * j)) {
        largest = i * j;
      }
    }
  }

  print(largest);
}